res/res_agi.c

```diff
@@ -1027,157 +1027,162 @@ static struct agi_cmd *get_agi_cmd(struct ast_channel *chan)
-
-		ast_debug(3, "STREAM %d (SSRC: %u) -- len: %ld samples: %d datalength: %d seqno: %d timestamp: %0.4f ts: %ld\n", stream_no, themssrc, f->len, f->samples, f->datalen, f->seqno, (float)c_ts/1000.00, c_ts);
-
-	       	if (ts_diff >= (2*f_ptime)) { // Twice the ptime size because ts in the ast_frame is saved based on the f_ptime
-        	       	ast_log(LOG_WARNING, "STREAM %d (SSRC: %u) Seqno (%d - %d) -- GAP (Sec): %f\t No of Frames Lost: %ld\n", stream_no, themssrc, f->seqno, last_seq, (float)ts_diff/1000.0, (ts_diff/f_ptime)-1);
-
-			/*! Insert Silence - by pasing
-			ts_start == (f_no) Last packet's ts incremented by f_ptime i.e. start of the ts for the 1st silent frame
-			f_ptime: ptime for the RTP stream
-			ts_diff: Difference in the ts for the current packet and last packet
-			ts_end === (f->ts) Current packet's ts value specifed for end of loop */
-			insert_silence(chan, f, fs, stream_no, f_no, f_ptime, ts_diff, c_ts, themssrc);
-		}
-	} else {
-		ast_log(LOG_NOTICE, "Stream %d (SSRC: %u) -- (c_ts %ld - l_ts %ld) c_ts is either less than l_ts or diff is larger than 15 mins\n", stream_no, themssrc, c_ts, l_ts);
-	}
-
-	/*! Save the ts and sequence number for the next check */
+            if ((ssrc_change == 1) && (ast_tvcmp(ast_channel_get_rec_end_ts(chan, stream_no), ast_tv(0, 0)) > 0)){
+                s_tv = ast_channel_get_rec_end_ts(chan, stream_no);
+            }
+            
+            gap_ms = ast_tvdiff_ms(ast_tvnow(), s_tv);
+            
+            if ((gap_ms/f_ptime) > 2){ // Greater than twice the ptime
+                ast_log(LOG_WARNING, "Stream %d (SSRC: %u) delayed by %ld (ms)...\n", stream_no, themssrc, gap_ms);
+                ast_debug(1, "Stream %d (SSRC: %u) -- pkt_diff: %ld\t f->ts: %ld\t last_ts: %ld\n", stream_no, themssrc, gap_ms, f->ts, ast_channel_get_last_ts(chan, stream_no));
+                
+                /*!NOTE: For the initial stream delay we should not depend on the f->ts (as it can be any random value) so we use gap_ms to fill in silence */
+                insert_silence(chan, f, fs, stream_no, 0, f_ptime, gap_ms, themssrc);
+            } else {
+                ast_log(LOG_NOTICE, "No Delay on Stream %d (SSRC: %u) !!!\n", stream_no, themssrc);
+            }
+        }
+        
+        /*! Save the ts and sequence number for the next check */
         ast_channel_set_last_ts(chan, f->ts, stream_no);
         ast_channel_set_last_seq(chan, f->seqno, stream_no);
-        ast_channel_set_pkt_count(chan, stream_no);
-	ast_channel_set_rec_end_ts(chan, stream_no);
-	return 0;
+        ast_channel_set_rec_end_ts(chan, stream_no);
+        return 0;
+    }
+    
+    cs_pkt_count = ast_channel_get_pkt_count(chan, stream_no) + ast_channel_get_extra_pkt_count(chan, stream_no);
+    os_pkt_count = stream_no==1?(ast_channel_get_pkt_count(chan, 2) + ast_channel_get_extra_pkt_count(chan, 2)):(ast_channel_get_pkt_count(chan, 1) + ast_channel_get_extra_pkt_count(chan, 1));
```

Comment:
This should work fine as well - 
`os_pkt_count = ast_channel_get_pkt_count(chan, stream_no ^ 0x03) + ast_channel_get_extra_pkt_count(chan, stream_no ^ 0x03)`